pkg/log/log.go

```diff
@@ -35,18 +35,50 @@ package log
 
 import (
 	"context"
+	"sync"
+	"time"
 
 	"github.com/go-logr/logr"
 )
 
 // SetLogger sets a concrete logging implementation for all deferred Loggers.
 func SetLogger(l logr.Logger) {
+	loggerWasSetLock.Lock()
+	defer loggerWasSetLock.Unlock()
+
+	loggerWasSet = true
 	Log.Fulfill(l)
 }
 
+// It is safe to assume that if this wasn't set within the first 30 seconds of a binaries
+// lifetime, it will never get set. The DelegatingLogger causes a high number of memory
+// allocations when not given an actual Logger, so we set a NullLogger to avoid that.
+//
+// We need to keep the DelegatingLogger because we have various inits() that get a logger from
+// here. They will always get executed before any code that imports controller-runtime
+// has a chance to run and hence to set an actual logger.
+func init() {
+	// Init is blocking, so start a new goroutine
+	go func() {
+		time.Sleep(30 * time.Second)
+		loggerWasSetLock.Lock()
+		defer loggerWasSetLock.Unlock()
+		if !loggerWasSet {
+			Log.Fulfill(NullLogger{})
```

Comment:
Do we want to print out a warning or error of some sort before transitioning the logger?

It seems that the delegating logger default might not be the best choice if it could potentially fill up memory if used for a long period of time.

We could also change direction here and default to a well known logger instead?

Reply:
> Do we want to print out a warning or error of some sort before transitioning the logger?

Hm good idea, but what logger should we use? Generally, this doesn't change anything unless someone changed the logger after more than 30 seconds before which I think is unlikely

> It seems that the delegating logger default might not be the best choice if it could potentially fill up memory if used for a long period of time.

See the description in the PR, we need it because various packages construct a logger via `init` which always happens before any non-controller-runtime code is executed. Without the delegating logger it is impossible to set the logger for those. It is only inefficient if no logger is set.

> We could also change direction here and default to a well known logger instead?

Sure

Reply:
Yeah that makes sense, I was thinking when we're transitioning we could print an error to stderr via the stdlib log package

Reply:
The alternative is to just default to `klogr` and let folks use a different logger if they want to, we could do the init of the default logger early when the package loads up, which should solve the problem described above

Reply:
I don't think it properly solves the problem, because right now you can set the global base logger for everything via `SetLogger` (which IMHO is convenient and useful). What you describe would only allow to override the default logger within the components.

Also, c-r doesn't use klogr anywhere that I am aware, it uses `zap` with a `logr.Logger` wrapper.

So right now the whole thing is documented as not logging anything if `SetLogger` is not called, so changing that would make this PR break compatibility, hence I would prefer to stick with the NullLogger so this can be backported. WDYT?

Reply:
That sounds good for now yeah, I was more brainstorming for later / long temr